File: android/src/main/java/com/eugeniuszx/reactnativemediapicker/CameraPermission.kt
package com.eugeniuszx.reactnativemediapicker

/**
 * The camera permission status reported to JS. [value] is the string the JS union is made of.
 *
 * [NOT_REQUIRED] means the app does not declare `android.permission.CAMERA`, so the system camera
 * app is launched without any runtime permission. [UNAVAILABLE] means the device has no camera at
 * all — a device that has one but no app able to handle the capture intent still reports its real
 * permission status here, and surfaces as `camera_unavailable` from `launchCamera`.
 */
internal enum class CameraPermission(val value: String) {
  GRANTED("granted"),
  NOT_DETERMINED("not_determined"),
  DENIED("denied"),
  BLOCKED("blocked"),
  NOT_REQUIRED("not_required"),
  UNAVAILABLE("unavailable"),
  ;

  companion object {
    /**
     * [shouldShowRationale] alone cannot tell "never asked" from "permanently denied" — both report
     * false — so [hasAsked] is tracked by this library and consulted first.
     */
    fun resolve(
      hasCamera: Boolean,
      declared: Boolean,
      granted: Boolean,
      hasAsked: Boolean,
      shouldShowRationale: Boolean,
    ): CameraPermission = when {
      !hasCamera -> UNAVAILABLE
      granted -> GRANTED
      !declared -> NOT_REQUIRED
      !hasAsked -> NOT_DETERMINED
      shouldShowRationale -> DENIED
      else -> BLOCKED
    }
  }
}

File: ios/Core/CameraPermission.swift
import Foundation

/// The camera authorization states this library distinguishes, mirroring `AVAuthorizationStatus`
/// without depending on AVFoundation.
enum CameraAuthorization {
  case notDetermined
  case authorized
  case denied
  case restricted
  case unknown
}

/// The camera permission status reported to JS.
///
/// Two more values exist in the JS union and are Android-only: `not_required` (the app does not
/// declare `android.permission.CAMERA`) and `denied` (refused, but re-askable). iOS grants an app a
/// single chance to ask, so a refusal there is `blocked` — only Settings can undo it.
enum CameraPermission: String {
  case granted
  case notDetermined = "not_determined"
  case blocked
  case unavailable

  static func resolve(hasCamera: Bool, authorization: CameraAuthorization) -> CameraPermission {
    guard hasCamera else { return .unavailable }
    switch authorization {
    case .authorized: return .granted
    case .notDetermined: return .notDetermined
    case .denied, .restricted, .unknown: return .blocked
    }
  }
}
